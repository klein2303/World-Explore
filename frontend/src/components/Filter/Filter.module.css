.filtercontainer {
    background-color: #E4E1F7;
    height: fit-content;
    width: 180px;

    display: flex;
    justify-content: center;
    padding: 20px;
    border-radius: 10px;
}

.informationbox {
    height: fit-content;
    width: 100%;
}

.title {
    color: #424242;
    font-size: 22px;
}

.decription {
    color: #424242;
    font-size: 17px;
    margin-top: 10px;
}

.filterlist {
    font-size: 15px;
    gap: 14px;
}

.filteritem {
    display: flex;
    align-items: center;
    gap: 5px;  
    color: #424242;  
}

.checkbox {
    display: none;
}

.customCheckbox {
    cursor: pointer;
    display: flex;
    gap: 5px;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    color: #424242;
}

.checkboxVisual {
    height: 12px;
    width: 12px;
    border: 1px solid #424242;
    border-radius: 3px;
    display: flex;
    justify-content: center;
    align-items: center;
}

.checkbox:checked + label .checkboxVisual {
    background-color: #424242; 
}

@media (max-width: 620px) {
    .title {
        display: none;
    }

    .decription {
        display: none;
    }

    .filtercontainer {
        background-color: white;
        height: fit-content;
        width: 340px;
    
        display: flex;
        justify-content: center;
        margin-left: 10px;
    }

    .informationbox {
        height: fit-content;
        width: 100%;
    }

    .filterlist {
        font-size: 13px;
        gap: 14px;

        display: flex;
        flex-direction: row;
        
        overflow-x: auto;
        white-space: nowrap;
    }

    .checkboxVisual {
        display: none;
    }

    .customCheckbox {
        height: fit-content;
        width: fit-content;
        padding: 5px;

        border-radius: 10px;
        border: 2px solid #E4E1F7;
    }

    .checkbox:checked + .customCheckbox {
        background-color: #E4E1F7; 
    }
}