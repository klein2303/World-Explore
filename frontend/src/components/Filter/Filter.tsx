import {useEffect} from "react";
import styles from "./Filter.module.css";
import {useRecoilState} from "recoil";
import {filterAtom} from "../../atoms/FilterAtom";
import {FilterType} from "../../types/FilterType";
import {setFilters} from "../../utils/FilterStorage";

const Filter = () => {
    const [isClicked, setIsClicked] = useRecoilState<FilterType>(filterAtom);

    const handleClick = (e: React.ChangeEvent<HTMLInputElement>) => {
        const {id, checked} = e.target;

        setIsClicked((prevState) => ({
            ...prevState,
            continent: {
                ...prevState.continent,
                [id]: checked,
            },
        }));
    };

    useEffect(() => {
        setFilters(isClicked);
    }, [isClicked]);

    return (
        <main className={styles.filtercontainer} aria-label="Filter based on continents">
            <div className={styles.informationbox} role="region" aria-labelledby="filters-title">
                <p id="filters-title" className={styles.title}>
                    Filters
                </p>
                <p id="continents-description" className={styles.decription}>
                    Continents
                </p>
                <div className={styles.filterlist} role="group" aria-labelledby="continent-filter">
                    <div className={styles.filteritem}>
                        <input
                            type="checkbox"
                            id="Africa"
                            name="Africa"
                            value="Africa"
                            onChange={handleClick}
                            checked={isClicked.continent["Africa"]}
                            className={styles.checkbox}
                            aria-labelledby="label-africa"
                        />
                        <label htmlFor="Africa" id="label-africa" className={styles.customCheckbox}>
                            <span className={styles.checkboxVisual} role="presentation"></span>
                            Africa
                        </label>
                    </div>
                    <div className={styles.filteritem}>
                        <input
                            type="checkbox"
                            id="Asia"
                            name="Asia"
                            value="Asia"
                            onChange={handleClick}
                            checked={isClicked.continent["Asia"]}
                            className={styles.checkbox}
                            aria-labelledby="label-asia"
                        />
                        <label htmlFor="Asia" id="label-asia" className={styles.customCheckbox}>
                            <span className={styles.checkboxVisual} role="presentation"></span>
                            Asia
                        </label>
                    </div>
                    <div className={styles.filteritem}>
                        <input
                            type="checkbox"
                            id="Europe"
                            name="Europe"
                            value="Europe"
                            onChange={handleClick}
                            checked={isClicked.continent["Europe"]}
                            className={styles.checkbox}
                            aria-labelledby="label-europe"
                        />
                        <label htmlFor="Europe" id="label-europe" className={styles.customCheckbox}>
                            <span className={styles.checkboxVisual} role="presentation"></span>
                            Europe
                        </label>
                    </div>
                    <div className={styles.filteritem}>
                        <input
                            type="checkbox"
                            id="North America"
                            name="North America"
                            value="North America"
                            onChange={handleClick}
                            checked={isClicked.continent["North America"]}
                            className={styles.checkbox}
                            aria-labelledby="label-north-america"
                        />
                        <label htmlFor="North America" id="label-north-america" className={styles.customCheckbox}>
                            <span className={styles.checkboxVisual} role="presentation"></span>
                            North America
                        </label>
                    </div>
                    <div className={styles.filteritem}>
                        <input
                            type="checkbox"
                            id="Oceania"
                            name="Oceania"
                            value="Oceania"
                            onChange={handleClick}
                            checked={isClicked.continent["Oceania"]}
                            className={styles.checkbox}
                            aria-labelledby="label-oceania"
                        />
                        <label htmlFor="Oceania" id="label-oceania" className={styles.customCheckbox}>
                            <span className={styles.checkboxVisual} role="presentation"></span>
                            Oceania
                        </label>
                    </div>
                    <div className={styles.filteritem}>
                        <input
                            type="checkbox"
                            id="South America"
                            name="South America"
                            value="South America"
                            onChange={handleClick}
                            checked={isClicked.continent["South America"]}
                            className={styles.checkbox}
                            aria-labelledby="label-south-america"
                        />
                        <label htmlFor="South America" id="label-south-america" className={styles.customCheckbox}>
                            <span className={styles.checkboxVisual} role="presentation"></span>
                            South America
                        </label>
                    </div>
                </div>
            </div>
        </main>
    );
};

export default Filter;
