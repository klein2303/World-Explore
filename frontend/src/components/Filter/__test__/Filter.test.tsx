import { render, screen, fireEvent} from "@testing-library/react";
import "@testing-library/jest-dom";
import { describe, it, expect, beforeEach} from "vitest";
import { MemoryRouter } from "react-router-dom";
import Filter from "../Filter";

describe("Filter Component Tests (Desktop)", () => {
  it("renders the filter component correctly", () => {
   
    const { asFragment } = render(
      <MemoryRouter>
        <Filter />
      </MemoryRouter>
    );

    // Snapshot the initial render
    expect(asFragment()).toMatchSnapshot();

    // Check if the title and description are in the document
    expect(screen.getByText("Filters")).toBeInTheDocument();
    expect(screen.getByText("Continents")).toBeInTheDocument();

    // Check if all continent checkboxes are rendered
    expect(screen.getByLabelText("Africa")).toBeInTheDocument();
    expect(screen.getByLabelText("Asia")).toBeInTheDocument();
    expect(screen.getByLabelText("Europe")).toBeInTheDocument();
    expect(screen.getByLabelText("North America")).toBeInTheDocument();
    expect(screen.getByLabelText("Oceania")).toBeInTheDocument();
    expect(screen.getByLabelText("South America")).toBeInTheDocument();
  });

  it("handles checkbox clicks and updates state", () => {
    render(
      <MemoryRouter>
        <Filter />
      </MemoryRouter>
    );

    // Select the Africa checkbox and check if it's originally not checked
    const africaCheckbox = screen.getByLabelText("Africa");
    expect(africaCheckbox).not.toBeChecked(); 

    // Simulate a click event on the Africa checkbox and check if it's checked
    fireEvent.click(africaCheckbox);
    expect(africaCheckbox).toBeChecked(); 

    // Uncheck the Africa checkbox and check if it's unchecked
    fireEvent.click(africaCheckbox);
    expect(africaCheckbox).not.toBeChecked(); 
  });
});

describe("Filter Component Tests (Mobile)", () => {
  // Simulate a mobile screen size before each test
  beforeEach(() => {
    window.innerWidth = 375; 
    window.dispatchEvent(new Event("resize"));
  });

  it("renders the filter component correctly in mobile view", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Filter />
      </MemoryRouter>
    );

    // Snapshot the mobile render
    expect(asFragment()).toMatchSnapshot();

    // Check that the filter container is present
    expect(screen.getByText("Filters")).toBeInTheDocument();
    expect(screen.getByText("Continents")).toBeInTheDocument();

    // Check if all checkboxes are rendered in mobile view
    expect(screen.getByLabelText("Africa")).toBeInTheDocument();
    expect(screen.getByLabelText("Asia")).toBeInTheDocument();
    expect(screen.getByLabelText("Europe")).toBeInTheDocument();
    expect(screen.getByLabelText("North America")).toBeInTheDocument();
    expect(screen.getByLabelText("Oceania")).toBeInTheDocument();
    expect(screen.getByLabelText("South America")).toBeInTheDocument();
  });

  it("handles checkbox interactions in mobile view", () => {
    render(
      <MemoryRouter>
        <Filter />
      </MemoryRouter>
    );

    // Select the Oceania checkbox
    const oceaniaCheckbox = screen.getByLabelText("Oceania");
    expect(oceaniaCheckbox).not.toBeChecked(); // Initially not checked

    // Simulate a click event on the Oceania checkbox and check if it's checked
    fireEvent.click(oceaniaCheckbox);
    expect(oceaniaCheckbox).toBeChecked(); 

    // Uncheck the Oceania checkbox and check if it's unchecked
    fireEvent.click(oceaniaCheckbox);
    expect(oceaniaCheckbox).not.toBeChecked(); 
  });
});
