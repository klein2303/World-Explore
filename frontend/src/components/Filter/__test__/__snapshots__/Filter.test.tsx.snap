// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filter Component Tests (Desktop) > renders the filter component correctly 1`] = `
<DocumentFragment>
  <main
    aria-description="Filter based on continents"
    class="_filtercontainer_435ebf"
  >
    <div
      aria-labelledby="filters-title"
      class="_informationbox_435ebf"
      role="region"
    >
      <p
        class="_title_435ebf"
        id="filters-title"
      >
        Filters
      </p>
      <p
        class="_decription_435ebf"
        id="continents-description"
      >
        Continents
      </p>
      <div
        aria-label="continents"
        class="_filterlist_435ebf"
        role="group"
      >
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-africa"
            class="_checkbox_435ebf"
            id="Africa"
            name="Africa"
            tabindex="0"
            type="checkbox"
            value="Africa"
          />
          <label
            class="_customCheckbox_435ebf"
            for="Africa"
            id="label-africa"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            Africa
          </label>
        </section>
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-asia"
            class="_checkbox_435ebf"
            id="Asia"
            name="Asia"
            tabindex="0"
            type="checkbox"
            value="Asia"
          />
          <label
            class="_customCheckbox_435ebf"
            for="Asia"
            id="label-asia"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            Asia
          </label>
        </section>
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-europe"
            class="_checkbox_435ebf"
            id="Europe"
            name="Europe"
            tabindex="0"
            type="checkbox"
            value="Europe"
          />
          <label
            class="_customCheckbox_435ebf"
            for="Europe"
            id="label-europe"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            Europe
          </label>
        </section>
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-north-america"
            class="_checkbox_435ebf"
            id="North America"
            name="North America"
            tabindex="0"
            type="checkbox"
            value="North America"
          />
          <label
            class="_customCheckbox_435ebf"
            for="North America"
            id="label-north-america"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            North America
          </label>
        </section>
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-oceania"
            class="_checkbox_435ebf"
            id="Oceania"
            name="Oceania"
            tabindex="0"
            type="checkbox"
            value="Oceania"
          />
          <label
            class="_customCheckbox_435ebf"
            for="Oceania"
            id="label-oceania"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            Oceania
          </label>
        </section>
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-south-america"
            class="_checkbox_435ebf"
            id="South America"
            name="South America"
            tabindex="0"
            type="checkbox"
            value="South America"
          />
          <label
            class="_customCheckbox_435ebf"
            for="South America"
            id="label-south-america"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            South America
          </label>
        </section>
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Filter Component Tests (Mobile) > renders the filter component correctly in mobile view 1`] = `
<DocumentFragment>
  <main
    aria-description="Filter based on continents"
    class="_filtercontainer_435ebf"
  >
    <div
      aria-labelledby="filters-title"
      class="_informationbox_435ebf"
      role="region"
    >
      <p
        class="_title_435ebf"
        id="filters-title"
      >
        Filters
      </p>
      <p
        class="_decription_435ebf"
        id="continents-description"
      >
        Continents
      </p>
      <div
        aria-label="continents"
        class="_filterlist_435ebf"
        role="group"
      >
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-africa"
            class="_checkbox_435ebf"
            id="Africa"
            name="Africa"
            tabindex="0"
            type="checkbox"
            value="Africa"
          />
          <label
            class="_customCheckbox_435ebf"
            for="Africa"
            id="label-africa"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            Africa
          </label>
        </section>
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-asia"
            class="_checkbox_435ebf"
            id="Asia"
            name="Asia"
            tabindex="0"
            type="checkbox"
            value="Asia"
          />
          <label
            class="_customCheckbox_435ebf"
            for="Asia"
            id="label-asia"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            Asia
          </label>
        </section>
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-europe"
            class="_checkbox_435ebf"
            id="Europe"
            name="Europe"
            tabindex="0"
            type="checkbox"
            value="Europe"
          />
          <label
            class="_customCheckbox_435ebf"
            for="Europe"
            id="label-europe"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            Europe
          </label>
        </section>
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-north-america"
            class="_checkbox_435ebf"
            id="North America"
            name="North America"
            tabindex="0"
            type="checkbox"
            value="North America"
          />
          <label
            class="_customCheckbox_435ebf"
            for="North America"
            id="label-north-america"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            North America
          </label>
        </section>
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-oceania"
            class="_checkbox_435ebf"
            id="Oceania"
            name="Oceania"
            tabindex="0"
            type="checkbox"
            value="Oceania"
          />
          <label
            class="_customCheckbox_435ebf"
            for="Oceania"
            id="label-oceania"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            Oceania
          </label>
        </section>
        <section
          class="_filteritem_435ebf"
        >
          <input
            aria-label="label-south-america"
            class="_checkbox_435ebf"
            id="South America"
            name="South America"
            tabindex="0"
            type="checkbox"
            value="South America"
          />
          <label
            class="_customCheckbox_435ebf"
            for="South America"
            id="label-south-america"
            tabindex="0"
          >
            <span
              class="_checkboxVisual_435ebf"
              role="presentation"
            />
            South America
          </label>
        </section>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
